alf/utils/value_ops.py

```diff
@@ -143,43 +138,39 @@ def discounted_return(rewards, values, step_types, discounts, time_major=True):
         A tensor with shape [T-1, B] (or [T-1]) representing the discounted
         returns. Shape is [B, T-1] when time_major is false.
     """
-
     if not time_major:
-        discounts = tf.transpose(a=discounts)
-        rewards = tf.transpose(a=rewards)
-        values = tf.transpose(a=values)
-        step_types = tf.transpose(a=step_types)
+        discounts = discounts.transpose(0, 1)
+        rewards = rewards.transpose(0, 1)
+        values = values.transpose(0, 1)
+        step_types = step_types.transpose(0, 1)
 
-    tf.Assert(
-        tf.shape(values)[0] >= 2, [
-            "The sequence length needs to be at least 2. Got ",
-            tf.shape(values)[0]
-        ])
+    assert values.shape[0] >= 2, "The sequence length needs to be \
+                                at least 2. Got {s}".format(s=values.shape[0])
 
+    returns = torch.zeros(rewards.shape, dtype=rewards.dtype)
     discounts = discounts[1:]
     rewards = rewards[1:]
     final_value = values[-1]
     values = values[:-1]
 
     step_types = step_types[:-1]
-    is_lasts = tf.cast(tf.equal(step_types, StepType.LAST), tf.float32)
+    is_lasts = (step_types == StepType.LAST).to(dtype=torch.float32)
 
-    def discounted_return_fn(acc_discounted_reward, args):
-        (reward, value, is_last, discount) = args
+    def discounted_return_fn(acc_discounted_reward, reward, value, is_last,
+                             discount):
         acc_discounted_value = acc_discounted_reward * discount + reward
         return is_last * value + (1 - is_last) * acc_discounted_value
 
-    returns = tf.scan(
-        fn=discounted_return_fn,
-        elems=(rewards, values, is_lasts, discounts),
-        reverse=True,
-        initializer=final_value,
-        back_prop=False)
+    returns[-1] = final_value
```

Comment:
You can avoid line 169 by this:
```python
returns = torch.zeros(rewards.shape, dtype=rewards.dtype)
ret = final_value
for t in reversed(range(rewards.shape[0])):
    ret = discounted_return_fn(returns[t + 1], rewards[t],
         values[t], is_lasts[t], discounts[t])
    returns[t] = ret
```

Reply:
Changed. Also did similar changes to normalized advantage estimation.